components/zcash_note_encryption/src/batch.rs

```diff
@@ -11,39 +10,50 @@ use crate::{
 /// Trial decryption of a batch of notes with a set of recipients.
 ///
 /// This is the batched version of [`crate::try_note_decryption`].
+///
+/// Returns a vector containing the decrypted result for each output,
+/// with the same length and in the same order as the outputs were
+/// provided, along with the index in the `ivks` slice associated with
+/// the IVK that successfully decrypted the output.
 #[allow(clippy::type_complexity)]
 pub fn try_note_decryption<D: BatchDomain, Output: ShieldedOutput<D, ENC_CIPHERTEXT_SIZE>>(
     ivks: &[D::IncomingViewingKey],
     outputs: &[(D, Output)],
-) -> Vec<Option<(D::Note, D::Recipient, D::Memo)>> {
+) -> Vec<Option<((D::Note, D::Recipient, D::Memo), usize)>> {
```

Comment:
Why make this a nested tuple, instead of just extending the existing tuple? Call sites will need to be updated either way.
```suggestion
) -> Vec<Option<(D::Note, D::Recipient, D::Memo, usize)>> {
```

Reply:
The reason this doesn't work is that the arity of the first element of this tuple varies according to the `decrypt_inner` argument to `batch_note_decryption`, and so we can't destructure it internally and reconstruct; it would take a whole extra allocation of the result vector and a needless pass over the result.